### DISCUSSION TITLE
CFP: Simplify masquerade configurations

### DISCUSSION THREAD

**Discussion by joestringer** [cilium/cilium#36335] [Ideas]
We have several overlapping options in Cilium for configuring masquerade based on interfaces, IP prefixes, routes, whether or not iptables is used for the implementation, and support for ip masq agent compatible configmaps.

All of these overlapping configurations can make it difficult for users to understand how they should configure Cilium in order to provide connectivity between workloads with private IP address space to the world. Anecdotally one of the most common reported problems from users is that connectivity is broken due to misconfigured masquerade settings.

I'd like to gather use cases for how masquerade is used in order to hopefully consolidate the options, set sane defaults, and provide easier documentation to users about masquerade settings if they do still need to manually adjust them.

This search filter may help in gathering information: https://github.com/cilium/cilium/issues?q=+label%3Afeature%2Fsnat+

**Comment by joestringer** [comment]
Simpler defaults :thread: 

As of `main` during v1.17 dev cycle:

```
$ git grep -i -e masq -e nativerouting daemon/cmd/daemon_main.go | grep flags
daemon/cmd/daemon_main.go:      flags.String(option.MasqueradeInterfaces, "", "Limit iptables-based egress masquerading to interfaces selector")
daemon/cmd/daemon_main.go:      flags.String(option.IPv4NativeRoutingCIDR, "", "Allows to explicitly specify the IPv4 CIDR for native routing. "+
daemon/cmd/daemon_main.go:      flags.String(option.IPv6NativeRoutingCIDR, "", "Allows to explicitly specify the IPv6 CIDR for native routing. "+
daemon/cmd/daemon_main.go:      flags.Bool(option.EnableIPv4Masquerade, true, "Masquerade IPv4 traffic from endpoints leaving the host")
daemon/cmd/daemon_main.go:      flags.Bool(option.EnableIPv6Masquerade, true, "Masquerade IPv6 traffic from endpoints leaving the host")
daemon/cmd/daemon_main.go:      flags.Bool(option.EnableBPFMasquerade, false, "Masquerade packets from endpoints leaving the host with BPF instead of iptables")
daemon/cmd/daemon_main.go:      flags.Bool(option.EnableMasqueradeRouteSource, false, "Masquerade packets to the source IP provided from the routing layer rather than interface address")
daemon/cmd/daemon_main.go:      flags.Bool(option.EnableIPMasqAgent, false, "Enable BPF ip-masq-agent")
daemon/cmd/daemon_main.go:      flags.String(option.IPMasqAgentConfigPath, "/etc/config/ip-masq-agent", "ip-masq-agent configuration file path")
daemon/cmd/daemon_main.go:      flags.Bool(option.InstallIptRules, true, "Install base iptables rules for cilium to mainly interact with kube-proxy (and masquerading)")
```

Masquerading is enabled for IPv4 and IPv6 by default, using the iptables-based mode. With https://github.com/cilium/cilium/pull/36103 , we could configure patterns for `ens+` and `eth+` by default to enable masquerading for common device naming conventions across the most common Linux distributions.

If the user does not configure the native routing CIDR explicitly, then for the most part the default seems to masquerade traffic from the local node's IPAM range destined for IPs outside the local node's IPAM range. This should work for most common cases for the default cluster-scope [IPAM mode](https://docs.cilium.io/en/v1.16/network/concepts/ipam/#address-management).

**Comment by joestringer** [comment]
Documentation :thread: 

The primary documentation for this feature is here: https://docs.cilium.io/en/stable/network/concepts/masquerading/#concepts-masquerading 

I'm seeing a couple of areas that could help a bit further on top of the current configuration:
- Highlight better that the ip masq approach supports multiple IP prefixes for exclusion from masquerade (maybe ensure this solution is in a dedicated section, with tradeoffs)
- Highlight the proper way to configure masquerade if using IPAM modes which don't allocate IP prefixes for each node (for example, cloud integrations with a global IP pool)
- Expand the `Configuration` section to give clearer guidance to users between the options, maybe include the routing-based option there (as a power user option)

**Comment by joestringer** [comment]
Use cases :thread:

**Reply by joestringer** [reply]
Just noting that https://github.com/cilium/cilium/pull/40132 proposes to introduce a use case where the decision about whether to masquerade traffic or not is dependent on source IP addresses of Pods, related to the Multi-Pool IPAM ranges that are configured in the environment.

**Reply by kevinvalk** [reply]
### Use case when running in native routing with eBPF masquerading when a (sub)set of the nodes do not have a shared network.

As discussed in #17177 (from august 2021), it might happen that Cilium is deployed in such a way that InternalIPs ARE natively routable but ExternalIPs are not (I think this is a pretty standard interpretation of the ExternalIP field).

However, Cilium assigns both the InternalIP and ExternalIPs the `reserved:remode-node` identity (6). This in turn "forces" no matter what you do, that **masquerading** is disabled for both ExternalIPs and InternalIPs. Even when using [`ipMasqAgent.enabled=true`](https://docs.cilium.io/en/stable/network/concepts/masquerading/#ebpf-based) and ensuring only the InternalIPs are on the `nonMasqueradeCIDRs`, all traffic to identities `reserved:remote-node` are still **NOT** masqueraded.

This is particularly a problem when pods try to access the `ExternalIPs`. For example, **cert-manager** does a sanity check (cannot be disabled) for HTTP challenges and tries to visit the URL to the .well-known challenge. If you are not using some kind of split DNS setup, this will always resolve to the ExternalIP addresses (how else can the website be served) and fail. Essentially, breaking **cert-manager** completely. 

#### Potential solution

There are two paths, either change/configure the assigned identity of ExternalIPs OR to allow for a more fine tune masquerade settings (for example via ipMasqAgent).

However, both approaches will be breaking and I can totally see how there are many Cilium deployments were the ExternalIPs ARE natively routable.

In my deployment I added a daemon that on node changes or after Cilium agent boots, changes the assigned identity of all ExternalIPs via `cilium-dbg bpf ipcache update`. While this works, it is quite some work and it feels rather fragile.

In my opinion, a more robust and extendable configuration would be to allow configuring how and what identities are assigned. This way you could even provide a different identity to ExternalIPs that can be used in network policies.

**Reply by joestringer** [reply]
Thanks for sharing. If this is mainly about masquerading and how the traffic forwarding works for traffic destined to an ExternalIP, then I'd rather take this the other way - decouple the Security Identity from the masquerade check altogether. Security Identity is intended as a policy primitive, not as a routing or NAT primitive. If Cilium is currently making a routing/NAT decision based on security identity, it is probably just because it's making an assumption about the topology of the network as it relates to different node IP addresses. That assumption feels like the things that is causing these problems.

It sounds like currently Cilium assumes that all remote nodes are accessible and have reverse routes to allow reply traffic to reach back to the source, but for some reason this is not the case in your environment. For that case, it could help for Cilium to have a better way to recognize the underlying routability of the ranges and then decide how to route to those peers, whether via a tunnel, masquerading, or directly routing without traffic modification.

**Comment by joestringer** [comment]
Cloning [this comment over from another PR](https://github.com/cilium/cilium/pull/40132#discussion_r2266795628) discussion via @alimehrabikoshki:

> By default, traffic leaving the cluster is masqueraded (enable-ipv4/6-masquerade).
> 
> We can opt-out of masquerading based on:
> 
> 1. Destination IP
> ipv4|6-native-routing-cidr or enable-ip-masq-agent - these set destination CIDRs that should not be masqueraded
> 2. Source IP
> New snat-exclusion-flags - if source IP is in the given CIDR, it will not masquerade.
> We could match either or both and things would work as expected, as they both disable masquerading, so no conflict there.
> 
> Exact masquerade behaviour is further customized by:
>
> 1. Interfaces
> egress-masquerade-interfaces - this limits masquerading to given interfaces (takes precedence over devices). This is compatible with snat-exclusion since they both try to limit when masquerading happens.
> 2. Routes
> enable-masquerade-to-route-source=true uses the routing layer to select source IP to masquerade, but does not make decisions on whether to masq or not. This could be used in conjunction with snat-exclusion since snat-exclusion comes first and enable-masquerade-to-route-source only applies to masqueraded traffic.
> 3. Datapath
to use BPF or not, via enable-bpf-masquerade. We support both in this MR.
> So the way I see it, these flags can be dropped in and configured with any other flag as needed.
